formats/json/jvmMain/src/kotlinx/serialization/json/internal/JvmJsonStreams.kt

```diff
@@ -0,0 +1,266 @@
+                        sz += escapedString.length
+                    }
+                    else -> {
+                        array[sz] = '\\'
+                        array[sz + 1] = marker.toInt().toChar()
+                        sz += 2
+                    }
+                }
+            } else {
+                array[sz++] = ch.toChar()
+            }
+        }
+        ensureTotalCapacity(sz, 1)
+        array[sz++] = '"'
+        writeUtf8(array, 0, sz)
+        flush()
+    }
+
+    private fun ensureTotalCapacity(oldSize: Int, additional: Int): Int {
+        val newSize = oldSize + additional
+        if (array.size <= newSize) {
+            array = array.copyOf(newSize.coerceAtLeast(oldSize * 2))
+        }
+        return oldSize
+    }
+
+    override fun release() {
+        flush()
+        CharArrayPool.release(array)
+        ByteArrayPool.release(buffer)
+    }
+
+    private fun flush() {
+        stream.write(buffer, 0, index)
+        index = 0
+    }
+
+
+    @Suppress("NOTHING_TO_INLINE")
+    // ! you should never ask for more than the buffer size
+    private inline fun ensure(bytesCount: Int) {
+        if (rest() < bytesCount) {
+            flush()
+        }
+    }
+
+    @Suppress("NOTHING_TO_INLINE")
+    // ! you should never ask for more than the buffer size
+    private inline fun write(byte: Int) {
+        buffer[index++] = byte.toByte()
+    }
+
+    @Suppress("NOTHING_TO_INLINE")
+    private inline fun rest(): Int {
+        return buffer.size - index
+    }
+
+    private fun writeUtf8(string: CharArray, beginIndex: Int, endIndex: Int) {
+        require(beginIndex >= 0) { "beginIndex < 0: $beginIndex" }
+        require(endIndex >= beginIndex) { "endIndex < beginIndex: $endIndex < $beginIndex" }
+        require(endIndex <= string.size) { "endIndex > string.length: $endIndex > ${string.size}" }
+
+        // Transcode a UTF-16 Java String to UTF-8 bytes.
+        var i = beginIndex
+        while (i < endIndex) {
+            var c = string[i].code
+
+            when {
+                c < 0x80 -> {
+                    // Emit a 7-bit character with 1 byte.
+                    ensure(1)
+                    write(c) // 0xxxxxxx
+                    i++
+                    val runLimit = minOf(endIndex, i + rest())
+
+                    // Fast-path contiguous runs of ASCII characters. This is ugly, but yields a ~4x performance
+                    // improvement over independent calls to writeByte().
+                    while (i < runLimit) {
+                        c = string[i].code
+                        if (c >= 0x80) break
+                        write(c) // 0xxxxxxx
+                        i++
+                    }
+                }
+
+                c < 0x800 -> {
+                    // Emit a 11-bit character with 2 bytes.
+                    ensure(2)
+                    /* ktlint-disable no-multi-spaces */
```

Comment:
We don't need these ktlint comments since we don't use it

Reply:
Also, it may be a gesture of good will to mention the source of the code (it's from okio afair?) in the comments, even though we are not obliged to do this by license